token/confidential-transfer/elgamal-registry/src/instruction.rs

```diff
@@ -0,0 +1,196 @@
+            }
+        };
+        buf
+    }
+}
+
+/// Create a `RegistryInstruction::CreateRegistry` instruction
+pub fn create_registry(
+    funding_address: &Pubkey,
+    owner_address: &Pubkey,
+    proof_location: ProofLocation<PubkeyValidityProofData>,
+) -> Result<Vec<Instruction>, ProgramError> {
+    let elgamal_registry_address = get_elgamal_registry_address(owner_address, &id());
+
+    let mut accounts = vec![
+        AccountMeta::new(*funding_address, true),
+        AccountMeta::new(elgamal_registry_address, false),
+        AccountMeta::new_readonly(*owner_address, false),
+        AccountMeta::new_readonly(system_program::id(), false),
+    ];
+    let proof_instruction_offset = proof_instruction_offset(&mut accounts, proof_location);
+
+    let registry_instruction = Instruction {
+        program_id: id(),
+        accounts,
+        data: RegistryInstruction::CreateRegistry {
+            proof_instruction_offset,
+        }
+        .pack(),
+    };
+    append_zk_elgamal_proof(registry_instruction, proof_location)
+}
+
+/// Create a `RegistryInstruction::UpdateRegistry` instruction
+pub fn update_registry(
+    registry_account: &Pubkey,
+    owner: &Pubkey,
+    proof_location: ProofLocation<PubkeyValidityProofData>,
+) -> Result<Vec<Instruction>, ProgramError> {
+    let mut accounts = vec![AccountMeta::new(*registry_account, false)];
+    let proof_instruction_offset = proof_instruction_offset(&mut accounts, proof_location);
+    accounts.push(AccountMeta::new_readonly(*owner, true));
+
+    let registry_instruction = Instruction {
+        program_id: id(),
+        accounts,
+        data: RegistryInstruction::UpdateRegistry {
+            proof_instruction_offset,
+        }
+        .pack(),
+    };
+    append_zk_elgamal_proof(registry_instruction, proof_location)
+}
+
+/// Takes a `ProofLocation`, updates the list of accounts, and returns a
+/// suitable proof location
+fn proof_instruction_offset(
+    accounts: &mut Vec<AccountMeta>,
+    proof_location: ProofLocation<PubkeyValidityProofData>,
+) -> i8 {
+    match proof_location {
+        ProofLocation::InstructionOffset(proof_instruction_offset, proof_data) => {
+            accounts.push(AccountMeta::new_readonly(sysvar::instructions::id(), false));
+            if let ProofData::RecordAccount(record_address, _) = proof_data {
+                accounts.push(AccountMeta::new_readonly(*record_address, false));
+            }
+            proof_instruction_offset.into()
+        }
+        ProofLocation::ContextStateAccount(context_state_account) => {
+            accounts.push(AccountMeta::new_readonly(*context_state_account, false));
+            0
+        }
+    }
+}
+
+/// Takes a `RegistryInstruction` and appends the pubkey validity proof
+/// instruction
+fn append_zk_elgamal_proof(
+    registry_instruction: Instruction,
+    proof_data_location: ProofLocation<PubkeyValidityProofData>,
+) -> Result<Vec<Instruction>, ProgramError> {
```

Comment:
nit: the function doesn't use the registry instruction, so how about this just takes a `&mut Vec<Instruction>` and optionally appends the proof instruction?

Reply:
Good idea! Maybe this function can be refactored into the `proof-generation` crate and shared by both registry and token22 instruction constructors, but that is probably outside the scope of this PR.